### PR TITLE
Apollo3 Add support for SPI

### PR DESCRIPTION
**Pull request by alistair23**
### Pull Request Overview

Add support for the Apollo3 IOM SPI

### Testing Strategy

Tested by running the tests and monitoring the traffic with a logic analyser

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

File: chips/apollo3/src/iom.rs [hunk 1]
```diff
@@ -839,6 +1008,384 @@ impl<'a> hil::i2c::SMBusMaster for Iom<'a> {
...
+
+        Ok(())
+    }
+
+    fn set_rate(&self, rate: u32) -> Result<u32, ErrorCode> {
+        if self.op.get() != Operation::SPI {
+            return Err(ErrorCode::BUSY);
+        }
+
+        let div: u32 = 48000000 / rate; // Really should be 48000000_u32.div_ceil(rate)
```
Comment (twilfredo) [inline chips/apollo3/src/iom.rs:1271]:
```suggestion
        let div: u32 = (48000000 + rate - 1) / rate; // Change to `48000000_u32.div_ceil(rate)` when api out of nightly
```

File: chips/apollo3/src/iom.rs [hunk 2]
```diff
@@ -472,75 +517,172 @@ impl<'a> Iom<'_> {
...
+                        let idx = self.write_index.get();
+                        let data = u32::from_le_bytes(
+                            write_buffer[idx..(idx + 4)].try_into().unwrap_or([0; 4]),
+                        );
+
+                        self.registers.fifopush.set(data);
+                        self.write_index.set(idx + 4);
+                    }
+
+                    // Get an remaining data that isn't 4 bytes long
```
Comment (twilfredo) [inline chips/apollo3/src/iom.rs:658]:
```suggestion
                    // Get remaining data that isn't 4 bytes long
```

File: boards/redboard_artemis_nano/src/tests/spi_controller.rs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    debug!("[SPI] Setup iom0 transfers... ");
+    run_kernel_op(100);
+    spi_host.set_client(cb);
+    cb.reset();
+
+    let tx = cb.tx_data.take().unwrap();
+    let rx = cb.rx_data.take().unwrap();
+    cb.tx_len.set(tx.len());
+
+    //Set iom0 Configs
```
Comment (twilfredo) [inline boards/redboard_artemis_nano/src/tests/spi_controller.rs:190]:
```suggestion
    // Set iom0 Configs
```

File: boards/redboard_artemis_nano/src/tests/spi_controller.rs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    run_kernel_op(100);
+
+    spi_host.set_client(cb);
+    cb.reset();
+
+    let tx = cb.tx_data.take().unwrap();
+    let rx = cb.rx_data.take().unwrap();
+    cb.tx_len.set(tx.len());
+
+    //Set iom0 Configs
```
Comment (twilfredo) [inline boards/redboard_artemis_nano/src/tests/spi_controller.rs:148]:
```suggestion
    // Set iom0 Configs
```

**Review by twilfredo — DISMISSED** [review]

File: chips/apollo3/src/iom.rs [hunk 5]
```diff
@@ -2,12 +2,15 @@
 
 use core::cell::Cell;
```
Comment (bradjc) [inline chips/apollo3/src/iom.rs:3]: Update header comment

**Review by bradjc — APPROVED** [review]

**Review by twilfredo — APPROVED** [review]

**Review by bradjc — APPROVED** [review]

**Pull request merged**